core/domain/suggestion_jobs_one_off.py

```diff
@@ -471,3 +475,74 @@ def reduce(key, values):
             yield (key, len(values))
         else:
             yield (key, values)
+
+
+class DeleteAllQuestionAndInvalidSuggestionsOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to delete all questions and invalid suggestion models in the
+    server.
+
+    This job deletes following suggestion and connected models:
+        - Deletes all the question suggestions as the old question suggestions
+          have invalid state schema version number.
+        - Delete all suggestions which does not have corresponding feedback
+          models.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [suggestion_models.GeneralSuggestionModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            yield ('SUGGESTION_ALREADY_DELETED', 1)
+            return
+
+        feedback_model = feedback_models.GeneralFeedbackThreadModel.get_by_id(
+            item.id)
+
+        if feedback_model is not None and item.suggestion_type != (
+                suggestion_models.SUGGESTION_TYPE_ADD_QUESTION):
+            yield ('SUGGESTION_IGNORED', 1)
+            return
+
+        model_keys = []
+        model_keys.append(datastore_services.Key(
+            suggestion_models.GeneralSuggestionModel, item.id))
+
+        if feedback_model is not None:
+            model_keys.append(datastore_services.Key(
+                feedback_models.GeneralFeedbackThreadModel, item.id))
+
+        for message in feedback_services.get_messages(item.id):
+            model_keys.append(
+                datastore_services.Key(
+                    feedback_models.GeneralFeedbackMessageModel, message.id))
+
+        reply_to_id_models = (
+            email_models.GeneralFeedbackEmailReplyToIdModel.get_by_thread_id(
+                item.id))
+
+        model_keys += [
+            datastore_services.Key(
+                email_models.GeneralFeedbackEmailReplyToIdModel,
+                reply_to_id_model.id)
+            for reply_to_id_model in reply_to_id_models]
+
+        model_keys += feedback_models.GeneralFeedbackThreadUserModel.query(
+            feedback_models.GeneralFeedbackThreadUserModel.thread_id == item.id
+        ).fetch(keys_only=True)
+
+        if item.target_type == feconf.ENTITY_TYPE_EXPLORATION:
+            model_keys.append(
+                datastore_services.Key(
+                    feedback_models.FeedbackAnalyticsModel, item.target_id))
+
+        transaction_services.run_in_transaction(
+            datastore_services.delete_multi, model_keys)
+        yield ('SUGGESTION_DELETED', 1)
```

Comment:
Do the updates for all of these models also run in transactions? Otherwise, I don't see how this will protect against concurrency failures/inconsistencies.

Reply:
Yeah, we run the [create_suggestion](https://github.com/oppia/oppia/blob/develop/core/controllers/suggestion.py#L103) method in translation.

Reply:
What about all of the other models being deleted here? We're deleting entities from several models. Do all of those models use transactions for creation?

Note that given the additional context below, this probably doesn't actually matter much since we're only running it on the test server (in which case running into datastore contention is pretty unlikely).

Reply:
create_suggestion function creates all these connected/related models and the create_suggestion function runs in translation in the controller layer. 

Also, I'm fine with removing `transaction_services.run_in_transaction` method. I had this one earlier as I was deleting multiple models at once in the reduce method but later on, I changed the plan and worked on deleting models related to one suggestion at a time in the map section.